### ISSUE TITLE
[Proposal] [.NET 8] Use `ExperimentalAttribute` instead of `ObsoleteAttribute` to indicate alpha APIs.

### ISSUE THREAD

**Issue by philliphoff** [dapr/dotnet-sdk#1219]
## Describe the proposal

Currently alpha APIs are decorated with the `ObsoleteAttribute` which causes the compiler to generate a warning/error when used. While the generated message is not entirely accurate, it *does* ensure users see and understand the risks of using a potentially unstable API.

.NET 6 introduced a new attribute, [RequiresPreviewFeaturesAttribute](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.versioning.requirespreviewfeaturesattribute?view=net-6.0), which serves a purpose similar to our use of `ObsoleteAttribute`. This generates a similar message (but from the Rosyln analyzer rather than the compiler itself), and requires a slightly different method to disable (via MSBuild property).  The main disadvantage is disabling applies to all uses of such decorated APIs, rather than individual uses.

Still, `RequiresPreviewFeaturesAttribute` feels like a more appropriate way to indicate alpha APIs.

> Another option is to create our own attribute and set of Roslyn analyzers that have our specific desired semantics; this could then be a start to additional work to help ensure best practices when using the SDK.

> Update: .NET 8 introduces `ExperimentalAttribute`, which is even *more* closely aligned with our use case.

**Comment by halspang** [comment]
@philliphoff - Would this be a breaking change if we switched them all over? Or is there a way to avoid that?

Or, is this proposal just to use it for new features and the old features will be `[Obsolete]` until they're no longer preview?

**Comment by philliphoff** [comment]
@halspang Yes, it would be breaking in that it's a different compiler/analyzer error to disable.  That said, I just discovered that .NET 8 includes a new [ExperimentalAttribute](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.codeanalysis.experimentalattribute?view=net-8.0) which seems even closer to what we want.  (`RequiresPreviewFeaturesAttribute` was originally targeted at APIs in .NET itself but would still have been better than `ObsoleteAttribute`.) I suggest we wait until .NET 6 support is dropped and *then* switch to `ExperimentalAttribute`, as a whole, so users needn't have to change how they suppress the warnings more than once.

**Comment by Neme12** [comment]
Oh, this isn't about an analyzer, my bad.

**Comment by WhitWaldo** [comment]
Closed with the completion of #1564
